### ISSUE TITLE
Get Timeouts

### ISSUE THREAD

**Issue by chriscummings120** [StackExchange/StackExchange.Redis#450]
We are seeing some very intermittent timeouts on GetString operations. Our app is single threaded and is not generating much load on the Redis Server.

I took at peek at the Redis source code and there are a few places where the code eats exceptions (try catch{}). Of particular concern are the places where the code is executing on the thread pool. With the exception eaten I can see a scenario where the response could be lost with no exception seen.

Is anyone else seeing this?

**Comment by niemyjski** [comment]
I'm seeing timeouts when getting all keys (less than 10 keys) against a local redis 3.2 instance...

**Comment by NickCraver** [comment]
Are you two seeing exceptions in the latest 1.1.608 library? The task continuation changes were quite a change in 1.1.605 and later.

**Comment by NickCraver** [comment]
From @dj-curls in #481

> System.TimeoutException: Timeout performing GET commdevice.logging-3671, inst: 2903, queue: 3063, qu: 0, qs: 3062, qc: 1, wr: 1, wq: 1, in: 40886, ar: 0, clientName: IP-AC80020E, IOCP: (Busy=748,Free=4252,Min=2,Max=5000), WORKER: (Busy=339,Free=4661,Min=2,Max=5000), Local-CPU: 0%

Seeing a lot of timeouts on start up of our application. This is from a production server.
Server is running on an adequate (as far as I can see) AWS Elasticache instance.
Was using version 1.1.603, have upgraded to 1.1.605 - doesn't seem to have made a difference. 

Connection strings are

> private static string serversRead = $"{ ConfigurationManager.AppSettings.Get("RedisServer") },syncTimeout=1000,abortConnect=false,allowAdmin=true";

and

> private static string serversWrite = $"{ ConfigurationManager.AppSettings.Get("RedisServer") },syncTimeout=5000,abortConnect=false,allowAdmin=true";

'commdevice.logging-3671' is a particularly small piece of data, however there are larger values in there, but nothing too big.

I have attached the 'info' command output from the redis server
[redis-info.txt](https://github.com/StackExchange/StackExchange.Redis/files/467568/redis-info.txt)

Where can I find out what 'inst', 'queue', 'qu', 'qs', 'qc', 'wr', 'wq', 'in' and 'ar' refer to?
What can I do to start troubleshooting this?
Where can I look for any other info?

**Comment by NickCraver** [comment]
From @drake69 in #430 

Randomly I get this error:

> Timeout performing SET NHibernate-Cache:Medium:ACompany.FMS.Data.Objects.OperationStock.OilStockOperationBalance#77110:lock, inst: 0, mgr: ProcessReadQueue, err: never, queue: 0, qu: 0, qs: 0, qc: 0, wr: 0, wq: 0, in: 0, ar: 1, clientName: MIL-S-613, IOCP: (Busy=0,Free=1000,Min=24,Max=1000), WORKER: (Busy=0,Free=32767,Min=24,Max=32767), Local-CPU: 0% (Please take a look at this article for some common client-side issues that can cause timeouts: https://github.com/StackExchange/StackExchange.Redis/tree/master/Docs/Timeouts.md)

Reading the suggested article I didn't find what's happening, can somebody point me to the right direction to resolve o workaround this error, I'm using a Hibernate Second Level Cache which is based on the main library StackExchange.Redis

**Comment by NickCraver** [comment]
From @tamnm in #341:

We using StackExchange.Redis v1.0.488 in Amazon Cloud server
and sometime it throw exception
> Timeout performing GET {KEY}, inst: 1, mgr: ProcessReadQueue, err: never, queue: 3, qu: 0, qs: 3, qc: 0, wr: 0, wq: 0, in: 7, ar: 1, IOCP: (Busy=13,Free=987,Min=4,Max=1000), WORKER: (Busy=12,Free=2035,Min=4,Max=2047)

**Comment by dj-curls** [comment]
Hi, 

I have since updated to the latest version v1.1.608 and I do still see timeouts. Is there anything else I can try, or anywhere I can look for more information as to why this is happening?

Thanks 
Oliver

**Comment by Tazer** [comment]
```
Timeout performing EVAL, inst: 1, mgr: ProcessReadQueue, err: never, queue: 569, qu: 0, qs: 569, qc: 0, wr: 0, wq: 0, in: 65536, ar: 1, clientName: server, IOCP:(Busy=3,Free=997,Min=2,Max=1000), WORKER:(Busy=14,Free=32753,Min=2,Max=32767), Local-CPU: unavailable (Please take a look at this article for some common client-side issues that can cause timeouts: https://github.com/StackExchange/StackExchange.Redis/tree/master/Docs/Timeouts.md)
```

We are seeing same problems with CacheManager.net and redis as a backplate. Running on Amazon against elastic cache (redis)

**Comment by nirajramshrestha** [comment]
Hi @NickCraver  and @mgravell ,

I am trying to follow up the issue posted by sachinNetweb. We are still having this issue intermittently in our system. Do you have any suggestion or idea why are we getting timeouts?

Exception we are getting currently,
Server:OSAPP01; 
An exception of type System.TimeoutException occurred and was caught.
Source:StackExchange.Redis;
Message:Timeout performing EXISTS OS:882:dataretrievaltime, inst: 1, mgr: ExecuteSelect, err: never, queue: 2, qu: 0, qs: 2, qc: 0, wr: 0, wq: 0, in: 0, ar: 0, clientName: ABC, IOCP: (Busy=0,Free=1000,Min=2,Max=1000), WORKER: (Busy=3,Free=32764,Min=2,Max=32767), Local-CPU: unavailable (Please take a look at this article for some common client-side issues that can cause timeouts: https://github.com/StackExchange/StackExchange.Redis/tree/master/Docs/Timeouts.md);
Target:T ExecuteSyncImpl[T](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1[T], StackExchange.Redis.ServerEndPoint);

StackTrace:   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server)
   at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server)
   at StackExchange.Redis.RedisDatabase.KeyExists(RedisKey key, CommandFlags flags)
   at ePlus.OneSource.StackExchangeRedis.StackExchangeRedisCache.RedisGetString(String key)
 Error in RedisGetString(key=System.String key, );Redis Error Detail: 10.18.53.64:6369,abortConnect=False

**Comment by nmathur** [comment]
We are also getting these issues on MGET, EVAL etc .. Our website us hosted on azure .. both web and redis are on same region (Central India) .. 

Timeout performing MGET, inst: 1, mgr: Inactive, err: never, queue: 2, qu: 0, qs: 2, qc: 0, wr: 0, wq: 0, in: 55341, ar: 0, clientName: RD000D3AF00AEC, serverEndpoint: Unspecified/redis-server:port, keyHashSlot: -2, IOCP: (Busy=0,Free=1000,Min=2,Max=1000), WORKER: (Busy=2,Free=32765,Min=2,Max=32767), Local-CPU: unavailable (Please take a look at this article for some common client-side issues that can cause timeouts: https://github.com/StackExchange/StackExchange.Redis/tree/master/Docs/Timeouts.md)

I reviewed many posts here and some of them suggesting to work on host's performance. We worked a lot on performance improvisation but still setting these issues .. and now when are users are increase day by the the issue is also occurring more often .. 

our redis version is  1.2.1.0 .. azure is suggesting these solutions .. 
https://docs.microsoft.com/en-us/azure/redis-cache/cache-faq#what-are-some-of-the-considerations-when-using-common-redis-commands

But I am looking for some examples .. please help and share which type of keys can perform better and what can cause time out ..

**Comment by NickCraver** [comment]
For those hitting this, a lot of issues were fixed by the build in 1.2.6 and above - please upgrade and let us know if you're still hitting issues. Just closing this issue out to consolidate the timeout issues and clarifying what's current.
